### PR TITLE
Fix preserving layer orders with offline editing

### PR DESCRIPTION
**Pull request by m-kuhn**
This completely moves layer order management to core (in particular also the `hasCustomLayerOrder` flag).

There is a new class `QgsLayerTree` which is used as layer tree root. This class takes authority for the layer order with a read-only property `QList<QgsMapLayer*> layerOrder` and r/w properties `QList<QgsMapLayer*> customLayerOrder` and `bool hasCustomLayerOrder`.

This is one more step in the direction of getting rid of the custom server project parser and it encapsulates the management of layer order completely in the layer tree.

Test failures are expected.

### CODE REVIEW

File: src/core/layertree/qgslayertree.h [hunk 1]
```diff
@@ -26,40 +26,201 @@
...
+     */
+    void hasCustomLayerOrderChanged( bool hasCustomLayerOrder );
+
+  private slots:
+    void nodeAddedChildren( QgsLayerTreeNode *node, int indexFrom, int indexTo );
+    void nodeRemovedChildren();
+
+  private:
+    void addMissingLayers();
+    QgsWeakMapLayerPointerList mCustomLayerOrder;
```
Comment (nyalldawson) [inline src/core/layertree/qgslayertree.h]: Missing an initializer here

Comment (nyalldawson) [inline src/core/layertree/qgslayertree.h]: (for mCustomLayerOrder)

Comment (nyalldawson) [inline src/core/layertree/qgslayertree.h]: Sorry, mHasCustomLayerOrder

**Review by nyalldawson — APPROVED** [review]
Good approach - I like it!

File: src/core/layertree/qgslayertree.h (hunk 1 above)
Comment (m-kuhn) [inline src/core/layertree/qgslayertree.h]: Done, thanks

**Comment by m-kuhn** [comment]
Merged

**Comment by 3nids** [comment]
Awesome, thanks!

**Pull request closed without merging**